aten/src/ATen/cuda/CUDAStream.cpp

```diff
@@ -1,57 +1,127 @@
+  static std::deque<std::atomic<int>> high_priority_counters;
+  static std::vector<std::vector<CUDAStreamInternals>> low_priority_streams;
+  static std::vector<std::vector<CUDAStreamInternals>> high_priority_streams;
+
+  // Thread-local current streams
   static thread_local CUDAStreamInternals** current_streams = nullptr;
 
-  // Creates a(n indestructible) default stream for each device
-  // Note: the default stream on each device is signified by a zero
-  // value for the pointer, and so is not actually created as usual.
+  // Populates global values and creates a default stream for each device.
+  // Note: the default stream on each device is signified by a nullptr, 
+  // and so is not created as usual.
   // In particular, we don't need to switch devices when creating the
   // streams.
-  static void initDefaultCUDAStreams() {
+  static void initGlobalStreamState() {
     num_gpus = getNumGPUs();
-    default_streams = (CUDAStreamInternals*) malloc(num_gpus * sizeof(CUDAStreamInternals));
+
+    // Resizes deques and vectors
+    default_streams.resize(num_gpus);
+    device_flags.resize(num_gpus);
+    low_priority_counters.resize(num_gpus);
+    high_priority_counters.resize(num_gpus);
+    low_priority_streams.resize(num_gpus);
+    high_priority_streams.resize(num_gpus);
+
+    // Initializes streams and counters
     for (auto i = decltype(num_gpus){0}; i < num_gpus; ++i) {
-      default_streams[i].is_destructible = false;
-      default_streams[i].refcount = 0;
       default_streams[i].device = i;
-      default_streams[i].stream = DEFAULT_STREAM;
+      low_priority_counters[i] = 0;
+      high_priority_counters[i] = 0;
+    }
+
+    // Populates low and high priority range
+    #ifndef __HIP_PLATFORM_HCC__
+      AT_CUDA_CHECK(cudaDeviceGetStreamPriorityRange(&LOW_PRIORITY, &HIGH_PRIORITY));
+    #endif 
+  }
+
+  // Creates the low and high priority stream pools for the specified device
+  static void initDeviceStreamState(const int64_t device) {
+    // Switches to the requested device so streams are properly associated
+    // with it.
+    at::DeviceGuard device_guard{(int)device};
+
+    low_priority_streams[device].resize(STREAMS_PER_POOL);
+    high_priority_streams[device].resize(STREAMS_PER_POOL);
+    
+    for (auto i = decltype(STREAMS_PER_POOL){0}; i < STREAMS_PER_POOL; ++i) {
+      auto& lowpri_stream = low_priority_streams[device][i];
+      auto& hipri_stream = high_priority_streams[device][i];
+      
+      lowpri_stream.device = device;
+      hipri_stream.device = device;
+      
+      #ifndef __HIP_PLATFORM_HCC__
+        AT_CUDA_CHECK(cudaStreamCreateWithPriority(
+          &lowpri_stream.stream
+        , DEFAULT_FLAGS
+        , LOW_PRIORITY));
+        AT_CUDA_CHECK(cudaStreamCreateWithPriority(
+          &hipri_stream.stream
+        , DEFAULT_FLAGS
+        , HIGH_PRIORITY));
+      #else 
+        AT_CUDA_CHECK(cudaStreamCreateWithFlags(
```

Comment:
Maybe a comment here with something like, "HIP doesn't support stream priority, so we just create two streams with the same priority."